File: server/src/main/java/org/codi/catan/model/response/GameResponse.java
/*
 * @author the-codinator
 * created on 2020/5/26
 */

package org.codi.catan.model.response;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.Getter;
import lombok.Setter;
import org.codi.catan.model.core.StrongEntity;
import org.codi.catan.model.game.Board;

@Getter
public class GameResponse implements StrongEntity {

    public GameResponse(Board board, StateResponse state) {
        this.id = board.getId();
        this.board = board;
        this.state = state;
        this.eTag = state.getETag();
    }

    private final String id;
    private final Board board;
    private final StateResponse state;
    @SuppressWarnings("checkstyle:MemberName")
    @Setter
    private String eTag;

    @Override
    @JsonIgnore
    public String getETag() {
        return this.eTag;
    }
}

File: server-azure/src/model/response/game-response.ts
import type { IdentifiableEntity, StrongEntity } from '../core';
import type { Board } from '../game/board';
import type { StateResponse } from './state-response';
import type { Writable } from 'ts-essentials';

export interface GameResponse extends IdentifiableEntity, StrongEntity {
  board: BoardResponse;
  state: StateResponse;
}

export type BoardResponse = Board;

export function createGameResponse(board: BoardResponse, state: StateResponse): GameResponse {
  const etag = state.etag;
  delete (state as Writable<StateResponse>).etag;
  return { id: board.id, board, state, etag };
}
